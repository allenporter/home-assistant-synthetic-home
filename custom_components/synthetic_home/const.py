"""Constants for Synthetic Home."""

DOMAIN = "synthetic_home"
DEFAULT_NAME = "Synthetic Home"
CONF_FILENAME = "config_filename"

# Used for loading initial state of an entity and it persists across config
# entry reloads. Updated with service calls.
DATA_DEVICE_STATES = "device_states"

ATTR_AREA_NAME = "area"
ATTR_DEVICE_NAME = "device"
ATTR_DEVICE_STATE_KEY = "device_state_key"
ATTR_CONFIG_ENTRY_ID = "config_entry_id"
