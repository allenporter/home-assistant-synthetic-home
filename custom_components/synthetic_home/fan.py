"""Fan platform for Synthetic Home."""

import logging
from typing import Any

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.components.fan import (
    FanEntity,
    FanEntityFeature,
    DOMAIN as FAN_DOMAIN,
)

from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import DOMAIN
from .entity import SyntheticEntity
from .model import ParsedEntity, filter_attributes

_LOGGER = logging.getLogger(__name__)

SUPPORTED_ATTRIBUTES = set(
    {
        "supported_features",
        "device_class",
        "is_on",
        "oscillating",
        "current_direction",
        "percentage",
        "preset_mode",
        "preset_modes",
        "speed_count",
    }
)


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_devices: AddEntitiesCallback
) -> None:
    """Set up fan platform."""
    synthetic_home = hass.data[DOMAIN][entry.entry_id]

    async_add_devices(
        SyntheticFan(
            entity,
            state=entity.state,
            **filter_attributes(entity, SUPPORTED_ATTRIBUTES),
        )
        for entity in synthetic_home.entities
        if entity.platform == FAN_DOMAIN
    )


class SyntheticFan(SyntheticEntity, FanEntity):
    """synthetic_home fan class."""

    _attr_reports_position = False

    def __init__(
        self,
        entity: ParsedEntity,
        state: str | None,
        *,
        supported_features: FanEntityFeature | None = None,
        is_on: bool | None = None,
        oscillating: bool | None = None,
        current_direction: str | None = None,
        percentage: int | None = None,
        preset_mode: str | None = None,
        preset_modes: list[str] | None = None,
        speed_count: int | None = None,
    ) -> None:
        """Initialize the SyntheticFan."""
        super().__init__(entity)
        if supported_features is not None:
            self._attr_supported_features = (
                FanEntityFeature(0)
                | FanEntityFeature.TURN_ON
                | FanEntityFeature.TURN_OFF
                | supported_features
            )
        if state is not None:
            self._attr_percentage = 100 if (state == "on") else 0
        elif is_on is not None:
            self._attr_percentage = 100 if is_on else 0
        if percentage is not None:
            self._attr_percentage = percentage
        if oscillating is not None:
            self._attr_oscillating = oscillating
        if current_direction is not None:
            self._attr_current_direction = current_direction
        if preset_mode is not None:
            self._attr_preset_mode = preset_mode
        if preset_modes is not None:
            self._preset_modes = preset_modes
        if speed_count is not None:
            self._attr_speed_count = speed_count
        _LOGGER.debug("self._attr_percentage=%s", self._attr_percentage)

    async def async_set_direction(self, direction: str) -> None:
        """Set the direction of the fan."""
        self._attr_current_direction = direction
        self.async_write_ha_state()

    async def async_set_preset_mode(self, preset_mode: str) -> None:
        """Set the preset mode of the fan."""
        self._attr_preset_mode = preset_mode
        self.async_write_ha_state()

    async def async_set_percentage(self, percentage: int) -> None:
        """Set the speed percentage of the fan."""
        self._attr_percentage = percentage
        self.async_write_ha_state()

    async def async_turn_on(
        self,
        percentage: int | None = None,
        preset_mode: str | None = None,
        **kwargs: Any,
    ) -> None:
        """Turn on the fan."""
        self._attr_percentage = 100
        self.async_write_ha_state()

    async def async_turn_off(self, **kwargs: Any) -> None:
        """Turn the fan off."""
        self._attr_percentage = 0
        self.async_write_ha_state()

    async def async_oscillate(self, oscillating: bool) -> None:
        """Oscillate the fan."""
        self._attr_oscillating = oscillating
        self.async_write_ha_state()
