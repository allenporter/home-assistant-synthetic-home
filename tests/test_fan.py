"""Test Synthetic Home fan."""

import pytest


from homeassistant.const import Platform
from homeassistant.components.fan import (
    DOMAIN as FAN_DOMAIN,
    SERVICE_OSCILLATE,
    ATTR_OSCILLATING,
)
from homeassistant.const import ATTR_ENTITY_ID, SERVICE_TURN_ON, SERVICE_TURN_OFF
from homeassistant.core import HomeAssistant

from .conftest import FIXTURES


@pytest.fixture(name="platforms")
def mock_platforms() -> list[Platform]:
    """Set up platform."""
    return [Platform.FAN]


@pytest.mark.parametrize(
    ("config_yaml_fixture", "test_entity"),
    [(f"{FIXTURES}/fan-example.yaml", "fan.counter_fan")],
)
async def test_fan(
    hass: HomeAssistant, setup_integration: None, test_entity: str
) -> None:
    """Test fan."""

    state = hass.states.get(test_entity)
    assert state
    assert state.state == "off"
    assert state.attributes == {
        "friendly_name": "Counter Fan",
        "supported_features": 3,
        "oscillating": False,
        "percentage": 0,
        "percentage_step": 1.0,
        "preset_mode": None,
        "preset_modes": None,
    }

    await hass.services.async_call(
        FAN_DOMAIN,
        SERVICE_TURN_ON,
        service_data={ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(test_entity)
    assert state
    assert state.state == "on"

    await hass.services.async_call(
        FAN_DOMAIN,
        SERVICE_OSCILLATE,
        service_data={ATTR_ENTITY_ID: test_entity, ATTR_OSCILLATING: True},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(test_entity)
    assert state
    assert state.state == "on"
    assert state.attributes.get("oscillating")

    await hass.services.async_call(
        FAN_DOMAIN,
        SERVICE_OSCILLATE,
        service_data={ATTR_ENTITY_ID: test_entity, ATTR_OSCILLATING: False},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(test_entity)
    assert state
    assert state.state == "on"
    assert not state.attributes.get("oscillating")

    await hass.services.async_call(
        FAN_DOMAIN,
        SERVICE_TURN_OFF,
        service_data={ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(test_entity)
    assert state
    assert state.state == "off"
