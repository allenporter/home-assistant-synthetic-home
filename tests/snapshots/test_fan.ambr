# serializer version: 1
# name: test_fan_device_state[tests/fixtures/fan-off.yaml]
  tuple(
    'off',
    ReadOnlyDict({
      'friendly_name': 'Counter Fan',
      'oscillating': False,
      'percentage': 0,
      'percentage_step': 1.0,
      'preset_mode': None,
      'preset_modes': None,
      'supported_features': <FanEntityFeature.SET_SPEED|OSCILLATE|TURN_OFF|TURN_ON: 51>,
    }),
  )
# ---
# name: test_fan_device_state[tests/fixtures/fan-on.yaml]
  tuple(
    'on',
    ReadOnlyDict({
      'friendly_name': 'Counter Fan',
      'oscillating': False,
      'percentage': 100,
      'percentage_step': 1.0,
      'preset_mode': None,
      'preset_modes': None,
      'supported_features': <FanEntityFeature.SET_SPEED|OSCILLATE|TURN_OFF|TURN_ON: 51>,
    }),
  )
# ---
# name: test_fan_device_state[tests/fixtures/fan-oscillating.yaml]
  tuple(
    'on',
    ReadOnlyDict({
      'friendly_name': 'Counter Fan',
      'oscillating': True,
      'percentage': 100,
      'percentage_step': 1.0,
      'preset_mode': None,
      'preset_modes': None,
      'supported_features': <FanEntityFeature.SET_SPEED|OSCILLATE|TURN_OFF|TURN_ON: 51>,
    }),
  )
# ---
